# Задание 7 Jupiter

## Подготовка данных

1. ### Загружаем датасет с файлами train.csv и test.csv как pandas.DataFrame

In [5]:
import pandas as pd
test_df = pd.read_csv('test.csv')
train_df = pd.read_csv('train.csv')

display(test_df.head())
display(train_df.head())

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

2. ### Конкатенируем данные в один DataFrame по ключу PassengerID

In [19]:
df = pd.concat([train_df, test_df], ignore_index=True)
df['Survived'] = df['Survived'].astype('category')
df['Pclass'] = df['Pclass'].astype('category')
df['Sex'] = df['Sex'].astype('category')

display(df)
display(df.info())

PassengerId Survived Pclass  \
0               1      0.0      3   
1               2      1.0      1   
2               3      1.0      3   
3               4      1.0      1   
4               5      0.0      3   
...           ...      ...    ...   
1304         1305      NaN      3   
1305         1306      NaN      1   
1306         1307      NaN      3   
1307         1308      NaN      3   
1308         1309      NaN      3   

                                                   Name     Sex   Age  SibSp  \
0                               Braund, Mr. Owen Harris    male  22.0      1   
1     Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                                Heikkinen, Miss. Laina  female  26.0      0   
3          Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                              Allen, Mr. William Henry    male  35.0      0   
...                                                 ...     ...   ...    ...   
1304                                 Spector, Mr. Woolf    male   NaN      0   
1305                       Oliva y Ocana, Dona. Fermina  female  39.0      0   
1306                       Saether, Mr. Simon Sivertsen    male  38.5      0   
1307                                Ware, Mr. Frederick    male   NaN      0   
1308                           Peter, Master. Michael J    male   NaN      1   

      Parch              Ticket      Fare Cabin Embarked  
0         0           A/5 21171    7.2500   NaN        S  
1         0            PC 17599   71.2833   C85        C  
2         0    STON/O2. 3101282    7.9250   NaN        S  
3         0              113803   53.1000  C123        S  
4         0              373450    8.0500   NaN        S  
...     ...                 ...       ...   ...      ...  
1304      0           A.5. 3236    8.0500   NaN        S  
1305      0            PC 17758  108.9000  C105        C  
1306      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
1307      0              359309    8.0500   NaN        S  
1308      1                2668   22.3583   NaN        C  

[1309 rows x 12 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   PassengerId  1309 non-null   int64   
 1   Survived     891 non-null    category
 2   Pclass       1309 non-null   category
 3   Name         1309 non-null   object  
 4   Sex          1309 non-null   category
 5   Age          1046 non-null   float64 
 6   SibSp        1309 non-null   int64   
 7   Parch        1309 non-null   int64   
 8   Ticket       1309 non-null   object  
 9   Fare         1308 non-null   float64 
 10  Cabin        295 non-null    object  
 11  Embarked     1307 non-null   object  
dtypes: category(3), float64(2), int64(3), object(4)
memory usage: 75.8+ KB


None

- Отсутствующие данные Nan
- Все столбцы корректных типов
- Выживание, класс и пол категориальные

## Анализ таблицы

1. ### Визуализируем базовую статистику таблицы

In [30]:
print(df.describe(include='all'))

max_pass = df['Pclass'].value_counts().idxmax()
print(f'\nВ классе {max_pass} наибольшее количество пассажиров')

        PassengerId  Survived  Pclass                  Name   Sex  \
count   1309.000000     891.0  1309.0                  1309  1309   
unique          NaN       2.0     3.0                  1307     2   
top             NaN       0.0     3.0  Connolly, Miss. Kate  male   
freq            NaN     549.0   709.0                     2   843   
mean     655.000000       NaN     NaN                   NaN   NaN   
std      378.020061       NaN     NaN                   NaN   NaN   
min        1.000000       NaN     NaN                   NaN   NaN   
25%      328.000000       NaN     NaN                   NaN   NaN   
50%      655.000000       NaN     NaN                   NaN   NaN   
75%      982.000000       NaN     NaN                   NaN   NaN   
max     1309.000000       NaN     NaN                   NaN   NaN   

                Age        SibSp        Parch    Ticket         Fare  \
count   1046.000000  1309.000000  1309.000000      1309  1308.000000   
unique          NaN        

### В каком классе было больше всего пассажиров?
Ответ: 3

2. ### Группировуем таблицу в два уровня: класс и пол, по среднему значению возраста

In [23]:
# group by class and sex
# count mean age
grouped = df.groupby(['Pclass', 'Sex'], observed=False)['Age'].mean()

display(grouped)

# find age gap
youngest = grouped.idxmin()
oldest = grouped.idxmax()
age_gap = grouped.max() - group.min()

print(f'Самая юный - класс {youngest[0]}, {youngest[1]} со средним возрастом {group.min():.2f}')
print(f'Самая взрослый - класс {oldest[0]}, {oldest[1]} со средним возрастом {group.max():.2f}')
print(f'Разница в возрасте - {age_gap:.2f}')

Pclass  Sex   
1       female    37.037594
        male      41.029272
2       female    27.499223
        male      30.815380
3       female    22.185329
        male      25.962264
Name: Age, dtype: float64

Самая юный - класс 3, female со средним возрастом 22.19
Самая взрослый - класс 1, male со средним возрастом 41.03
Разница в возрасте - 18.84


### Кто из возможных комбинаций самый юный, кто самый взрослый? 
Ответ: Самый юный - класс 3, female со средним возрастом 22.19,
        
        самый взрослый - класс 1, male со средним возрастом 41.03
### Насколько отличаются эти значения?
Ответ: Разница в возрасте - 18.84

3. ### Отбираем только выживших пассажиров с фамилией, начинающейся на “K”
   ### Отсортировываем их по убыванию стоимости билета

In [27]:
#  select the survivors with the last name on "K"
survivors_K = df[(df['Survived'] == 1) & (df['Name'].str.startswith('K'))]

# sort by descending ticket price
sorted_survivors = survivors_K.sort_values(by='Fare', ascending=False)

display(sorted_survivors.head())

# find the most expensive and cheapest ticket
most_expensive = sorted_survivors.iloc[0]
cheapest = sorted_survivors.iloc[-1]

print(f'Самый дорогой билет - {most_expensive["Fare"]}    у пассажира - {most_expensive["Name"]}')
print(f'Самый дешевый билет - {cheapest["Fare"]}    у пассажира {cheapest["Name"]}')

PassengerId Survived Pclass                                 Name     Sex  \
621          622      1.0      1         Kimball, Mr. Edwin Nelson Jr    male   
457          458      1.0      1    Kenyon, Mrs. Frederick R (Marion)  female   
316          317      1.0      2  Kantor, Mrs. Sinai (Miriam Sternin)  female   
184          185      1.0      3  Kink-Heilmann, Miss. Luise Gretchen  female   
706          707      1.0      2        Kelly, Mrs. Florence "Fannie"  female   

      Age  SibSp  Parch  Ticket     Fare Cabin Embarked  Relatives  
621  42.0      1      0   11753  52.5542   D19        S          1  
457   NaN      1      0   17464  51.8625   D21        S          1  
316  24.0      1      0  244367  26.0000   NaN        S          1  
184   4.0      0      2  315153  22.0250   NaN        S          2  
706  45.0      0      0  223596  13.5000   NaN        S          0

Самый дорогой билет - 52.5542    у пассажира - Kimball, Mr. Edwin Nelson Jr
Самый дешевый билет - 7.75    у пассажира Kelly, Miss. Mary

У выжившего пассажира Asplund, Mrs. Carl Oscar (Selma Augusta Emilia Johansson) было максимальное количество родных - 6


### Кто заплатил больше всех? Кто меньше всех?
Ответ: Kimball, Mr. Edwin Nelson Jr заплатил больше всех (52.5542),

        Kelly, Miss. Mary заплатила меньше всех (7.75)

In [28]:
# the max number of relatives
df['Relatives'] = df['SibSp'] + df['Parch']
idx_max_relatives = df[df['Survived'] == 1]['Relatives'].idxmax()

print(f"\nУ выжившего пассажира {df.iloc[idx_max_relatives]['Name']} "
      f"было максимальное количество родных - {df.iloc[idx_max_relatives]['Relatives']}")


У выжившего пассажира Asplund, Mrs. Carl Oscar (Selma Augusta Emilia Johansson) было максимальное количество родных - 6


4. ### Какое максимальное количество родных было с выжившим пассажиром?
Ответ: 6

5. ### Считаем среднюю стоимость билета пассажиров, для которых указана каюта (Cabin) и для тех, у кого она не указана

In [29]:
# count mean cost and difference
mean_with_cabins = df[df['Cabin'].notna()]['Fare'].mean()
mean_without_cabins = df[df['Cabin'].isna()]['Fare'].mean()
diff = mean_with_cabins/mean_without_cabins
print(f"Средняя стоимость билета пассажиров с каютой: {mean_with_cabins:.2f}")
print(f"Средняя стоимость билета пассажиров без каюты: {mean_without_cabins:.2f}")
print(f"Стоимость отличается в {diff:.2f} раз")

Средняя стоимость билета пассажиров с каютой: 81.93
Средняя стоимость билета пассажиров без каюты: 19.13
Стоимость отличается в 4.28 раз


Средняя стоимость билета пассажиров с каютой: 81.93

Средняя стоимость билета пассажиров без каюты: 19.13
### Во сколько раз они отличаются?
Ответ: В 4.28 раз